### DISCUSSION TITLE
Terraform-ls produces huge amount of log entries

### DISCUSSION THREAD

**Discussion by jpaju** [helix-editor/helix#9630] [Troubleshooting] [answered]
Hi!

When using terraform-ls with the default configuration, it produces a huge number (10-20k rows per minute) and fills the `helix.log` file. All the rows in the log file are like this:
`2024-02-14T18:04:15.178 helix_lsp::transport [ERROR] terraform-ls err <- "2024/02/14 18:04:15 rpc_logger.go:53: Response to \"textDocument/references\" (ID 6): []\n"`.

It seems that Helix interprets the responses from terraform-ls as errors 🤔 I'm not sure whether this is an issue with Helix, or terraform-ls not adhering to the LSP spec.

Is it possible to suppress the logging Helix does per LSP?  Any other solutions for this?

**Comment by the-mikedavis** [comment, marked as answer]
See https://github.com/helix-editor/helix/blob/4df08ddbe02f1e8d84c6aa4be810a91f83c73441/helix-lsp/src/transport.rs#L74

and

https://github.com/helix-editor/helix/blob/4df08ddbe02f1e8d84c6aa4be810a91f83c73441/helix-lsp/src/transport.rs#L139-L151

We log everything the server sends over its stderr as an error. Some language servers use stderr as a logging backend since it's simpler (and I suppose tidier) than having a dedicated log file. It looks like it's possible to configure a file to send the logs to instead of stderr and terraform-ls is looking into changing their logging strategy https://github.com/hashicorp/terraform-ls/issues/1271. I would recommend adding the arguments to redirect the logs to `/dev/null` (though I don't think that's crossplatform) or to a file of your choice. Ideally this would be solved upstream though

**Reply by jpaju** [reply]
Huge thanks for the lightning fast and accurate answer! 🙏

The explanation makes perfect sense and, like you suggested, I was able to stop flooding the logs with this config in `languages.toml`

```toml
[language-server.terraform-ls]
command = "terraform-ls"
args = ["serve", "-log-file", "/dev/null"]
```
